core/src/main/java/io/methvin/watcher/visitor/DefaultFilesVisitor.java

```diff
@@ -0,0 +1,48 @@
+package io.methvin.watcher.visitor;
+
+import java.io.IOException;
+import java.nio.file.FileVisitResult;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.SimpleFileVisitor;
+import java.nio.file.attribute.BasicFileAttributes;
+
+public class DefaultFilesVisitor implements FilesVisitor {
+  public DefaultFilesVisitor(FailedCallback failedCallback) {
+    this.failedCallback = failedCallback;
+  }
+
+  private final FailedCallback failedCallback;
+
+  @Override
+  public void recursiveVisitFiles(Path file, Callback onDirectory, Callback onFile)
+      throws IOException {
+    SimpleFileVisitor<Path> visitor =
+        new SimpleFileVisitor<>() {
+          @Override
+          public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs)
+              throws IOException {
+            onDirectory.call(dir);
+            return FileVisitResult.CONTINUE;
+          }
+
+          @Override
+          public FileVisitResult visitFile(Path file, BasicFileAttributes attrs)
+              throws IOException {
+            onFile.call(file);
+            return FileVisitResult.CONTINUE;
+          }
+
+          @Override
+          public FileVisitResult visitFileFailed(Path file, IOException exc) throws IOException {
```

Comment:
Do we also need to do the same thing for `postVisitDirectory`? I think that will get called with the exception in case there's an exception iterating over the directory that wasn't related to a specific file.